### ISSUE TITLE
Emulation Support?

### ISSUE THREAD

**Issue by Hedronmx** [tkashkin/GameHub#103]
Could you add support for emulators?

**Comment by tkashkin** [comment]
Right now it's possible to add custom game with `+` button in headerbar.
How do you think emulator support should be added?
* One source per emulator.
* One source for all emulated games.
* Use existing compatibility layers system.

**Comment by Hedronmx** [comment]
You could use libretro for this, use it as a source. Then the user could select which "Core" to use (A core is a game emulator, like snes9x or dolphin) and select the rom. 
Maybe you could use a core as a source but I believe that could be too cumbersome.

What do you think??

**Comment by tkashkin** [comment]
RetroArch is now supported as a compatibility layer.

Add game ROM as a custom game, force compatibility mode in its properties and select RetroArch as a compatibility layer. Then select appropriate core and run.

![image](https://user-images.githubusercontent.com/2003017/47271095-9e675300-d57d-11e8-92c9-75c1af318d72.png)

**Comment by tkashkin** [comment]
> How can I donate??

Thanks.
There were a discussion few days ago, but I have removed these comments as offtopic.

I have a couple of links in readme of `dev` branch, but I have problems with PayPal right now so I can't withdraw donations. Now my account is limited and will probably be banned soon.
I don't know any options besides PayPal. So it's probably impossible for me to accept donations.

It may be possible to send money directly to Visa card, but I doubt it will work.

**Comment by Hedronmx** [comment]
Alright, I'll wait for the donations.
I don't get a force compatibility layer button.. Is this normal?

**Comment by tkashkin** [comment]
You have probably encountered #99. I have no idea why that happens.

This issue is so weird. It works fine on locally built version, but works wrong even when I install deb package built on the same system.

Generated C code is identical, but for some reason it works differently.

**Comment by Hedronmx** [comment]
So... How can I fix this? Should I disable show non native games? Or should I build the app?

**Comment by tkashkin** [comment]
You can try to build app. It will help to understand if you have same behaviour.

```bash
git clone https://github.com/tkashkin/GameHub.git --single-branch -b dev --depth=1
meson build --prefix=/usr -Ddistro=debian --buildtype=debug
cd build
ninja && src/com.github.tkashkin.gamehub --debug
```

**Comment by Hedronmx** [comment]
Built the app, force compatibility button now shows...!

**Comment by tkashkin** [comment]
It doesn't wait for argument, it just does not correctly update when file is selected. If you select file and then change title, it will work. I'll fix that for the next release.

**Comment by Hedronmx** [comment]
Thank you so much! I love your support and app. Keep the good work!

**Comment by tkashkin** [comment]
Thanks.

Do you have any ideas about this weird behavior? I have no idea how can the same code work differently.
`valac` bug? `gcc` bug? CPU hardware bug? :confused:
I have checked C code generated by Vala compiler and it's equal (no diff and checksums are equal too).

**Comment by tkashkin** [comment]
I've got an idea. Package is built with optimizations (`--buildtype=debugoptimized`). My local version is using `--buildtype=debug`. I'll try to build a package without optimizations.

**Comment by tkashkin** [comment]
Yes, seems to be it. I'll update build scripts to disable it everywhere.

**Comment by Hedronmx** [comment]
Awesome! I'm waiting for the new package!

**Comment by tkashkin** [comment]
Hopefully should be fixed in 393c96e. Packages in releases and in PPA should be available soon.
https://ci.appveyor.com/project/tkashkin/gamehub/builds/19679395
https://launchpad.net/~tkashkin/+archive/ubuntu/gamehub/+builds?build_state=all

**Comment by neuromancer** [comment]
> Can we have a small Use this setup for every launch. I use your app with a controller and having to press launch everytime sucks a bit.

We already asked for [this to implement a 'big picture' mode](https://github.com/tkashkin/GameHub/issues/75#issue-360673231)

**Comment by Hedronmx** [comment]
@tkashkin Force compatibility button is not showing again... Updated from PPA

**Comment by tkashkin** [comment]
@Hedronmx

**Comment by tkashkin** [comment]
@Hedronmx 
* It still doesn't work in PPA.
* Latest version is not built yet, because GitHub is having problems now and webhooks are paused. You can build it yourself to test.

**Comment by Hedronmx** [comment]
Alright thank you! Sorry to bother. :D

**Comment by tkashkin** [comment]
It's now possible to add custom emulators. Emulators can be launched with compat tools, just like regular games.

![image](https://user-images.githubusercontent.com/2003017/47578258-a1818b00-d951-11e8-80c5-59c008dd78c2.png)

To use this you need to add emulator in settings, add custom game and force "Custom emulator" as a compatibility tool for this game. Then select emulator and run.

![image](https://user-images.githubusercontent.com/2003017/47578492-384e4780-d952-11e8-8a9c-6773df86c26c.png)

**Comment by tkashkin** [comment]
f5b3ad0 adds emulated games import.
![image](https://user-images.githubusercontent.com/2003017/59156661-ba394e00-8aa7-11e9-8c89-4821ae31dc38.png)
![image](https://user-images.githubusercontent.com/2003017/59156667-d210d200-8aa7-11e9-810a-6a0221368dd7.png)

GameHub parses RetroArch cores info for supported file extensions and scans directory with `find`. (Some cores should probably be blacklisted as they are not that useful)

Custom emulators now can have a game file pattern.
![image](https://user-images.githubusercontent.com/2003017/59156700-56635500-8aa8-11e9-8e0b-58a66a6ea4de.png)

**Comment by hlechner** [comment]
I was looking to find a way to filter my console game (emulated) from the rest of the games without needing to use tags, so, here my idea:

By creating a new platform filter: Linux, Windows, Mac and _**Emulators**_.

In this idea, everything that uses retroarch or custom emulator would receive the "Emulator" platform and loses the "Linux" platform.

Not sure if it's the best approach but at least it will be possible to filter the console games and also technically it's not native from Linux/Windows/Mac would make sense to not display as native from that these platforms.

**Comment by tkashkin** [comment]
I have added `Emulated` platform. You'll need to reimport emulated games to apply it.

**Comment by hlechner** [comment]
Awesome, working perfectly here

**Comment by CorporalQuesadilla** [comment]
Everything works perfectly! Only thing I've noticed is that Retroarch does not have a custom pattern option (at least, that I could find). As a result, .zip and .7z (among some other formats) are not picked up. So far these are the only two formats that are missing from my library, I'll keep you updated if I find more.

I'm not sure if the solution is just to add these extensions as found or to let the user have total control over which extensions are searched for.

I should mention that when it comes to file extraction, Retroarch already takes care of this on its own - there's no need to extract using GameHub and pass the result in - Retroarch can easily take in a zip (or in some (core-specific?) cases, 7z) and load only the appropriate file for its core.

**Comment by tkashkin** [comment]
@CorporalQuesadilla currently some cores and file extensions are ignored to not spam imported games list with possibly unrelated files:

https://github.com/tkashkin/GameHub/blob/480c20f0cf50f6d8e8407af62ffb5f39a96622a4/src/ui/dialogs/ImportEmulatedGamesDialog.vala#L36-L37

Maybe it should be configurable.

**Comment by CorporalQuesadilla** [comment]
That's fair, I can definitely see the benefits of having a blacklist, especially for .bin, which causes a lot of issues for users who should be using .cue in the first place. I think Genesis/MD sometimes use .bin for some ROMs, but overall the No-Intro standard is .md. Virtual Boy and DS also uses .bin sometimes. Still, bin is more likely to cause issues than not.

I've also noticed that PSX cores do accept .exe - an unlikely circumstance, though. I think that most Dosbox users would likely be using a standalone build instead of the libretro core, and agree that including exe by default would likely cause more clutter.

If we're going with the blacklist approach, perhaps there could be checkboxes for each so the user becomes aware of which extensions are ignored.

If we're going with a whitelist approach, I think all extensions should appear in a comma separated textbox, alongside a list of defaults (or a reset to default button).

Personally, I prefer the whitelist approach, as a complete blacklist would be rather difficult, and somewhat less useful since Linux users often have tons of files without extensions at all while ROMs are almost always packaged with some form of semi-unique extension.

For reference, here are the extensions recognized by most major cores:
`"smc", "sfc", "swc", "fig", "bs", "vb", "vboy", "gd3", "gd7", "dx2", "bsx", "gb", "gbc", "bml", "rom", "st", "min", "nds", "n64", "v64", "z64", "u1", "ndd", "3ds", "3dsx", "elf", "axf", "cci", "cxi", "app", "nes", "fds", "unf", "unif", "iso", "gcm", "dol", "tgc", "wbfs", "ciso", "gcz", "wad", "gba", "agb", "dmg", "sgb", "cgb", "fd", "sap", "k7", "m7", "m5", "ri", "mx1", "mx2", "col", "dsk", "cas", "sg", "sc", "m3u", "cue", "toc", "ccd", "pbp", "chd", "img", "mdf", "cbn", "cso", "prx", "tzx", "tap", "z80", "rzx", "scl", "trd", "p", "t81", "dim", "img", "d88", "88d", "hdm", "dup", "2hd", "xdf", "hdf", "cmd", "vms", "dci", "smd", "32x", "sms", "68k", "mdx", "gen", "gg", "sg", "md", "cdi", "gdi", "game", "88", "ngp", "ngc", "d98", "98d", "fdi", "fdd", "tfd", "hdi", "thd", "nhd", "hdd", "hdn", "pce", "sgx", "int", "vec", "ws", "wsc", "pc2", "a26", "a78", "xfd", "atr", "atx", "cdm", "a52", "xex", "j64", "jag", "abs", "cof", "prg", "lnx", "st", "msa", "stx", "dim", "ipf", "sna", "tap", "cdt", "voc", "kcr"`

There may be a few duplicates, but I went through every core and manually typed only the ones I felt worth including. No idea if it's any use to you, but I've been meaning to do this for my own personal projects anways. I did choose to leave out any archive formats (mostly zip and 7z) as well as anything associated with dosbox or otherwise super common formats.

**Comment by tkashkin** [comment]
Now blacklists are configurable in Settings/Emulators/RetroArch. Both lists should be separated by `|`.

**Comment by hlechner** [comment]
Question about the dosbox support: it should be displayed on Compatibility Layer?
![image](https://user-images.githubusercontent.com/11244067/60475461-ce5b1000-9c4d-11e9-9f59-4ea93511ae85.png)

**Comment by tkashkin** [comment]
@hlechner yes, DOSBox is displayed as compatibility layer if there is `dosbox` executable somewhere in `$PATH` and there are `.conf` files in game directory.

**Comment by hlechner** [comment]
I was running the games directly calling the dosbox without a `dosbox.conf` per game but using a global one on `~/.dosbox/dosbox.conf`. That's why DOSBox didn't show on the compatibility layer.

Maybe would be a good idea to document it then? I can create a wiki page if you want.

Based on [Arch Linux Wiki](https://wiki.archlinux.org/index.php/DOSBox) to create a file you need to run the command (inside the DOSBox):
`config -wc dosbox.conf`

And then you can manually edit the dosbox.conf to launch the game by adding the lines to execute the game on `[autoexec]` section, here the example of the game Pizza Tycoon:
```
MOUNT C "."
C:
PT.EXE
```

**Comment by tkashkin** [comment]
@hlechner

I don't think there's anything to improve here. Most DOSBox games have configs and GameHub should detect them.

Documentation would be nice though. I have created an [issue](https://github.com/tkashkin/GameHub/issues/270) for it.
Feel free to edit wiki pages if you want. Ask me if you have any questions.

**Comment by Sixdd6** [comment]
I've just found GameHub today and tried to import a Cemu game. I have set up Cemu as a custom emulator but when I go to import the game (Breath of the Wild) it fails to find the game. My folder structure is (Emulation/Games/Nintendo - WiiU/DECRYPTED/The Legend of Zelda Breath of the Wild [ALZE0101]/code/U-King.rpx) but no matter what folder I point GameHub to it just doesn't find the game at all. Also when try to add the rpx file directly it is greyed out and not selectable. So how would I go about importing exactly?

**Comment by tkashkin** [comment]
@Sixdd6 Have you configured Cemu as a custom emulator? To do this:

* Open Settings > Emulators
* Add an emulator, set its executable, directory and compatibility options.
* Set emulator arguments.
  For Cemu it should be something like `-g $file $game_args`.
  Add `-f` to emulator or game arguments to launch it in fullscreen.
* Set game file patterns:
  - **Executable** - primary file that needs to be passed to the emulator.
    For Cemu it should be something like `./code/*.rpx|*.wud|*.wux`
  - **Image, Icon** - GameHub can import images and icons if game bundles them in separate files somewhere in its directory.
    For Cemu you can use `./meta/bootTvTex.tga` as image and `./meta/iconTex.tga` as icon.

After you've configured an emulator, click `+` button in main window and Import emulated games.

**Comment by Sixdd6** [comment]
After checking through your instructions, yes the emulator is set up just fine. The problem is that GameHub does not detect the game properly, I click the + button and browse to where the .rpx file is at but the game is not detected. I have already tried running the game through Cemu through Wine and it works fine, GameHub simply does not detect the game.

**Comment by Sixdd6** [comment]
I found the problem, the values for executable and images were autofilled but not actually filled, they were slightly grey. Maybe a proper autofill for those values would be good since they disappear as soon as you click to input the exact same thing anyway.

**Comment by telanus** [comment]
When importing a few billion roms Gamehub seems to hang, as there is no indication that GameHub is importing anything. would be way more user friendly if there is any indication that GameHub is importing/working, since the current behaviour it looks a lot like nothing is happening and that GH has crashed
